File: pfe-java/app/controllers/Items.java
package controllers;

import controllers.oauth.OAuth;
import models.Item;
import play.data.Form;
import play.data.validation.Constraints;
import play.libs.Json;
import play.mvc.Result;

import javax.inject.Inject;
import javax.inject.Singleton;

import java.util.function.Function;

import static controllers.DBAction.DB;
import static controllers.Render.render;
import static controllers.Render.version;
import static play.mvc.Http.MimeTypes;

@Singleton
public class Items extends Controller {

    final OAuth oauth;

    @Inject
    public Items(Service service, OAuth oauth) {
        super(service);
        this.oauth = oauth;
    }

    public static class CreateItem {
        @Constraints.Required
        public String name;
        @Constraints.Required
        @Constraints.Min(0)
        public Double price;
    }

    @DB
    public Result list() {
        return render(
            version(MimeTypes.HTML, () -> ok(views.html.list.render(service.shop.list()))),
            version(MimeTypes.JSON, () -> ok(Json.toJson(service.shop.list())))
        );
    }

    public Result createForm() {
        return ok(views.html.createForm.render(Form.form(CreateItem.class)));
    }

    @DB
    public Result create() {
        Form<CreateItem> submission = Form.form(CreateItem.class).bindFromRequest();
        if (submission.hasErrors()) {
            return render(
                    version(MimeTypes.HTML, () -> badRequest(views.html.createForm.render(submission))),
                    version(MimeTypes.JSON, () -> badRequest(submission.errorsAsJson()))
            );
        } else {
            CreateItem createItem = submission.get();
            Item item = service.shop.create(createItem.name, createItem.price);
            if (item != null) {
                return render(
                        version(MimeTypes.HTML, () -> redirect(routes.Items.details(item.id))),
                        version(MimeTypes.JSON, () -> ok(Json.toJson(item)))
                );
            } else {
                return internalServerError();
            }
        }
    }

    @DB
    public Result details(Long id) {
        Item item = service.shop.get(id);
        if (item != null) {
            return render(
                    version(MimeTypes.HTML, () -> ok(views.html.details.render(item))),
                    version(MimeTypes.JSON, () -> ok(Json.toJson(item)))
            );
        } else {
            return notFound();
        }
    }

    @DB
    public Result update(Long id) {
        Form<CreateItem> submission = Form.form(CreateItem.class).bindFromRequest();
        if (submission.hasErrors()) {
            return badRequest(submission.errorsAsJson());
        } else {
            CreateItem updateItem = submission.get();
            Item updated = service.shop.update(id, updateItem.name, updateItem.price);
            if (updated != null) {
                return ok(Json.toJson(updated));
            } else {
                return internalServerError();
            }
        }
    }

    @DB
    public Result delete(Long id) {
        if (service.shop.delete(id)) {
            return ok();
        } else {
            return badRequest();
        }
    }

    public Result share(Long id) {
        return oauth.authenticated(token -> {
            service.socialNetwork.share(routes.Items.details(id).absoluteURL(request()), token);
            return ok();
        }, (() -> redirect(oauth.authorizeUrl(routes.Items.details(id)))));
    }

}

File: pfe-scala/app/controllers/Items.scala
package controllers

import controllers.oauth.OAuth
import play.api.Logger
import play.api.i18n.{MessagesApi, I18nSupport}
import play.api.libs.concurrent.Execution.Implicits.defaultContext
import play.api.mvc.{Controller, Action}
import play.api.libs.json._
import models.{SocialNetwork, Shop, Item}
import play.api.data.Form
import play.api.data.Forms.{mapping, text, of}
import play.api.data.format.Formats.doubleFormat
import play.api.data.validation.Constraints
import play.filters.csrf.CSRFAddToken

import scala.concurrent.Future

case class CreateItem(name: String, price: Double)

object CreateItem {
  val form = Form(mapping(
    "name" -> text.verifying(Constraints.nonEmpty),
    "price" -> of[Double].verifying(Constraints.min(0.0, strict = true))
  )(CreateItem.apply)(CreateItem.unapply))
}

class Items(shop: Shop, socialNetwork: SocialNetwork, oauth: OAuth, val messagesApi: MessagesApi) extends Controller with I18nSupport {

  import Items._

  val list = Action.async { implicit request =>
    shop.list().map { items =>
      render {
        case Accepts.Html() => Ok(views.html.list(items))
        case Accepts.Json() => Ok(Json.toJson(items))
      }
    }
  }

  val create = Action.async { implicit request =>
    CreateItem.form.bindFromRequest().fold(
      formWithErrors => Future.successful(
        render {
          case Accepts.Html() => BadRequest(views.html.createForm(formWithErrors))
          case Accepts.Json() => BadRequest(formWithErrors.errorsAsJson)
        }
      ),
      createItem => {
        shop.create(createItem.name, createItem.price).map { item =>
          render {
            case Accepts.Html() => Redirect(routes.Items.details(item.id))
            case Accepts.Json() => Ok(Json.toJson(item))
          }
        }
      }
    )
  }

  val createForm = Action { implicit request =>
    Ok(views.html.createForm(CreateItem.form))
  }

  def details(id: Long) = Action.async { implicit request =>
    shop.get(id).map {
      case Some(item) => render {
        case Accepts.Html() => Ok(views.html.details(item))
        case Accepts.Json() => Ok(Json.toJson(item))
      }
      case None => NotFound
    }
  }

  def update(id: Long) = Action.async { implicit request =>
    CreateItem.form.bindFromRequest().fold(
      formWithErrors => Future.successful(BadRequest(formWithErrors.errorsAsJson)),
      updateItem => shop.update(id, updateItem.name, updateItem.price).map {
        case Some(item) => Ok(Json.toJson(item))
        case None => InternalServerError
      }
    )
  }

  def delete(id: Long) = Action.async {
    shop.delete(id).map(deleted => if (deleted) Ok else BadRequest)
  }

  def share(id: Long) = Action { implicit request =>
    oauth.authenticated( token => {
      socialNetwork.share(routes.Items.details(id).absoluteURL(), token).foreach { response =>
        Logger.info(s"Sharing request successfully sent for item #$id. Got response status ${response.statusText}")
      }
      Ok
    }, Redirect(oauth.authorizeUrl(routes.Items.details(id))))
  }

}

object Items {

  implicit val writesItem: Writes[Item] = Json.writes[Item]

}